promql/engine_test.go

```diff
@@ -542,234 +711,496 @@ load 10s
+					Metric: lbls2,
+				},
+			},
+		}, { // Timestamps for matrix selector does not depend on the evaluation time.
+			query: "metric[20s] @ 300",
+			start: 10,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 28, T: 280000}, {V: 29, T: 290000}, {V: 30, T: 300000}},
+					Metric: lbls1,
+				},
+				Series{
+					Points: []Point{{V: 56, T: 280000}, {V: 58, T: 290000}, {V: 60, T: 300000}},
+					Metric: lbls2,
+				},
+			},
+		}, { // Timestamps for matrix selector does not depend on the evaluation time.
+			query: "metric[20s] @ 300",
+			start: 1000,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 28, T: 280000}, {V: 29, T: 290000}, {V: 30, T: 300000}},
+					Metric: lbls1,
+				},
+				Series{
+					Points: []Point{{V: 56, T: 280000}, {V: 58, T: 290000}, {V: 60, T: 300000}},
+					Metric: lbls2,
+				},
+			},
+		}, { // Timestamps for matrix selector does not depend on the evaluation time.
+			query: "metric[20s] @ 300",
+			start: -1000,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 28, T: 280000}, {V: 29, T: 290000}, {V: 30, T: 300000}},
+					Metric: lbls1,
+				},
+				Series{
+					Points: []Point{{V: 56, T: 280000}, {V: 58, T: 290000}, {V: 60, T: 300000}},
+					Metric: lbls2,
+				},
+			},
+		}, {
+			query: `metric_ms[3ms] @ 2.345`,
+			start: 100,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 2342, T: 2342}, {V: 2343, T: 2343}, {V: 2344, T: 2344}, {V: 2345, T: 2345}},
+					Metric: lblsms,
+				},
+			},
+		}, {
+			query: `metric_neg[2s] @ 0`,
+			start: 100,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 3, T: -2000}, {V: 2, T: -1000}, {V: 1, T: 0}},
+					Metric: lblsneg,
+				},
+			},
+		}, {
+			query: `metric_neg[2s] @ -0`,
+			start: 100,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 3, T: -2000}, {V: 2, T: -1000}, {V: 1, T: 0}},
+					Metric: lblsneg,
+				},
+			},
+		}, {
+			query: `metric_neg[3s] @ -500`,
+			start: 100,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 504, T: -503000}, {V: 503, T: -502000}, {V: 502, T: -501000}, {V: 501, T: -500000}},
+					Metric: lblsneg,
+				},
+			},
+		}, {
+			query: "metric[100s:25s] @ 300",
+			start: 100,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 20, T: 200000}, {V: 22, T: 225000}, {V: 25, T: 250000}, {V: 27, T: 275000}, {V: 30, T: 300000}},
+					Metric: lbls1,
+				},
+				Series{
+					Points: []Point{{V: 40, T: 200000}, {V: 44, T: 225000}, {V: 50, T: 250000}, {V: 54, T: 275000}, {V: 60, T: 300000}},
+					Metric: lbls2,
+				},
+			},
+		}, {
+			query: "metric[100s:25s] @ 50",
+			start: 100,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 0, T: 0}, {V: 2, T: 25000}, {V: 5, T: 50000}},
+					Metric: lbls1,
+				},
+				Series{
+					Points: []Point{{V: 0, T: 0}, {V: 4, T: 25000}, {V: 10, T: 50000}},
+					Metric: lbls2,
+				},
+			},
+		}, { // Different nested timestamps.
+			query: `sum_over_time(metric{job="1"}[100s] @ 100)[100s:25s] @ 50`,
```

Comment:
Keep what tests you do here to a minimum, many of these can go in the .test files. Any tests here should be either specific to potential failure modes of range queries or 1-2 very basic smoketests, not general testing of the feature.

Tests like these are much harder to read and write.

Reply:
I dont see a way to test range selection (range vector selectors and subqueries) in .test file, so I have them here. I will see how I can reduce them.